Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -1219,6 +1219,35 @@ static NavigationNavigationType determineNavigationType(FrameLoadType loadType,
     return NavigationNavigationType::Push;
 }
 
+// https://html.spec.whatwg.org/multipage/browsing-the-web.html#url-and-history-update-steps
```

Comment:
This needs a FIXME for integrating with History::stateObjectAdded.

Reply:
I'm not sure I understand what you mean.

Reply:
I mean that the code from History::stateObjectAdded is supposed to use the same algorithm:
https://html.spec.whatwg.org/multipage/nav-history-apis.html#shared-history-push/replace-state-steps

So basically, I think we want to avoid duplicating code that the code paths for history.pushState/replaceState and Navigation::innerDispatchNavigateEvent are doing, both ideally can just use the same algorithm (in FrameLoader::updateURLAndHistory).